README.md
# docker_igv_js [![Build Status](https://travis-ci.org/refinery-platform/docker_igv_js.svg?branch=master)](https://travis-ci.org/refinery-platform/docker_igv_js)

A Docker container to wrap IGV.js

## Motivation

Visualizations in [Refinery](https://github.com/refinery-platform/) are provided by Docker containers.
The link between Refinery and the Docker containers is [django_docker_engine](https://github.com/refinery-platform/django_docker_engine).
A container may provide back-end services to a visualization, or after start up it may just serve static html and js,
as is the case here. In either case the input files are provided in a directory which is mounted on a predetermined path.

## Add new genomes

The IGV instance will look for genome reference files in a particular S3 bucket. To add reference files for a new assembly,
check out [get-reference-genomes](https://github.com/refinery-platform/get-reference-genomes) and run `genome-to-s3.sh`.

## Development

Clone the repository, make sure Docker is installed, and then:

```
pip install -r requirements.txt
./build_run_test.sh
```

After the tests run the container is left up, with the local url where it can be accessed.

## Release

Successful Github tags and PRs will prompt Travis to push the built image to Dockerhub. For a new version number:

```
git tag v0.0.x && git push origin --tags
```


CHANGELOG.md
# Change Log

## [v0.0.9](https://github.com/refinery-platform/docker_igv_js/tree/v0.0.9) (Aug 13, 2018)

* Add CHANGELOG.md

## [v0.0.8](https://github.com/refinery-platform/docker_igv_js/tree/v0.0.8) (Aug 6, 2018)

* Allow input via either INPUT_JSON or INPUT_JSON_URL.

test.py
import docker
import requests
import sys
import time
import unittest

from test_utils import TestContainerRunner


class ContainerTest(unittest.TestCase):

    def get_url(self, name):
        client = docker.APIClient(base_url='unix://var/run/docker.sock')
        port = client.port(name, 80)[0]["HostPort"]
        url = 'http://localhost:{}'.format(port)

        for i in range(5):
            try:
                requests.get(url)
                return url
            except:
                print('Still waiting for server...')
                time.sleep(1)
        else:
            self.fail('Server never came up')

    def assert_expected_response(self, name, expected, path='/'):
        url = self.get_url(name)
        response = requests.get(url + path)
        # TODO: Not ideal for error pages
        self.assertEqual(200, response.status_code)
        self.assertIn(expected, response.text)

    # Good configurations:

    def test_good_home_page(self):
        self.assert_expected_response('good', '>IGV<')

    def test_input_data_url(self):
        self.assert_expected_response(
            'good',
            '{',
            path='/options.json'
        )

    # Bad configurations:

    def test_missing_assembly(self):
        self.assert_expected_response(
            'missing_assembly',
            'Unexpected 404 from https://s3.amazonaws.com/data.cloud.refinery-platform.org/data/igv-reference/hgFAKE/cytoBand.txt'
        )

    def test_multiple_assemblies(self):
        self.assert_expected_response(
            'multiple_assemblies',
            # If this happens often, could return more detail, but this is
            # enough, for now.
            'AssertionError()'
        )

    def test_no_parameters(self):
        self.assert_expected_response(
            'no_parameters',
            "KeyError('parameters',)"
        )


if __name__ == '__main__':
    with TestContainerRunner():
        suite = unittest.TestLoader().loadTestsFromTestCase(ContainerTest)
        result = unittest.TextTestRunner(verbosity=2).run(suite)

    if not result.wasSuccessful():
        sys.exit(1)


test_utils.py
import _thread
import docker
import os
import socket

from http.server import HTTPServer, SimpleHTTPRequestHandler


class TestFixtureServer(object):

    def __init__(self):
        self.port = 9999
        self.ip = self.get_python_server_ip()

    def get_python_server_ip(self):
        # https://stackoverflow.com/a/166589
        s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        s.connect(("8.8.8.8", 80))
        host_ip = s.getsockname()[0]
        s.close()
        return host_ip

    def _start_server(self):
        server = HTTPServer((self.ip, self.port), SimpleHTTPRequestHandler)
        server.serve_forever()

    def start_server_in_background(self):
        # start the server in a background thread
        _thread.start_new_thread(self._start_server, ())


class TestContainerRunner(object):

    def __init__(self):
        self.client = docker.from_env()
        self.image_name = "docker_igv_js"
        self.repository = "gehlenborglab/docker_igv_js"
        self.containers = []
        self.test_fixture_server = TestFixtureServer()
        self.test_fixture_server.start_server_in_background()

        self._pull_image()
        self._build_image()

    def __enter__(self):
        self.run()

    def __exit__(self, *args):
        if not os.environ.get("CONTINUOUS_INTEGRATION"):
            self.cleanup_containers()

    def _pull_image(self):
        self.client.images.pull(self.repository + ":latest")

    def _build_image(self):
        print("Building image: {}".format(self.image_name))
        self.client.images.build(
            path="context",
            tag=self.image_name,
            rm=True,
            forcerm=True,
            cache_from=[self.repository]
        )

    def run(self):
        print("Starting TestContainerRunner...")
        for test_fixture in os.listdir("input_fixtures"):
            print("Creating container: {}".format(test_fixture))
            container = self.client.containers.run(
                self.image_name,
                detach=True,
                name=test_fixture,
                environment={
                    "INPUT_JSON_URL":
                        "http://{}:{}/input_fixtures/{}/input.json".format(
                            self.test_fixture_server.ip,
                            self.test_fixture_server.port,
                            test_fixture
                        )
                },
                ports={"80/tcp": None},
                publish_all_ports=True,
                extra_hosts={socket.gethostname(): self.test_fixture_server.ip}
            )
            self.containers.append(container)

    def cleanup_containers(self):
        print("Cleaning up TestContainerRunner containers...")
        print(self.containers)
        for container in self.containers:
            container.remove(force=True, v=True)


build_run_test.sh
#!/usr/bin/env bash
set -e

source define_repo.sh

python test.py

context/on_startup.py
import SimpleHTTPServer
import SocketServer
import json
import os
import requests

from cgi import escape


def get_refinery_input():
    """ Make a GET request to acquire the input data for the container"""
    input_json_url = os.environ.get('INPUT_JSON_URL')
    if input_json_url:
        return requests.get(input_json_url).json()
    return json.loads(os.environ['INPUT_JSON'])


def write_igv_configuration():
    config_data = get_refinery_input()

    tracks = []
    assemblies = [
        parameter['value'] for parameter in config_data['parameters']
        if parameter['name'] == 'Genome Build'
    ]
    assert len(assemblies) == 1
    assembly = assemblies[0]

    url_base = "https://s3.amazonaws.com/data.cloud.refinery-platform.org/data/igv-reference/{}/".format(
        assembly
    )
    for node_data in config_data["node_info"].values():
        tracks.append(
            {
                "name": "{} - {}".format(
                    node_data["node_solr_info"]["name"],
                    node_data["file_url"]
                ),
                "url": node_data["file_url"]
            }
        )
    reference = {
        "fastaURL": url_base + assembly + ".fa",
        "indexURL": url_base + assembly + ".fa.fai",
        "cytobandURL": url_base + "cytoBand.txt",
    }

    validate_urls(reference.values())

    options = {
        "reference": reference,
        "tracks": tracks
    }
    with open('options.json', 'w') as options_file:
        options_file.write(json.dumps(options))


def validate_urls(urls):
    url_status = {}
    for url in urls:
        try:
            # byte-range so we don't download the file; S3 does not support
            # HEAD.
            status = requests.get(
                url,
                headers={'Range': 'bytes=0-0'}
            ).status_code
        except requests.exceptions.RequestException as e:
            status = e.message
        url_status[url] = status
    messages = [
        'Unexpected {} from {}'.format(status, url)
        for url, status in url_status.iteritems()
        if status != 206  # If byte-ranges are handled, should be 206, not 200.
    ]
    if messages:
        raise Exception('\n'.join(messages))


def start_server():
    Handler = SimpleHTTPServer.SimpleHTTPRequestHandler
    SocketServer.TCPServer(("", 80), Handler).serve_forever()


if __name__ == '__main__':
    try:
        write_igv_configuration()
    except Exception as e:
        html = '<html><body><pre>{}</pre></body></html>'.format(
            # Print whole error if message too short
            escape(e.message if ' ' in e.message else repr(e))
        )
        with open('index.html', 'w') as index_file:
            index_file.write(html)
    start_server()
